pkg/products/monitoringspec/reconciler.go

```diff
@@ -0,0 +1,388 @@
+
+	if serviceMonitor.Spec.NamespaceSelector.Any {
+		logrus.Warnf("servicemonitor : %s cannot be copied to %s namespace. Namespace selector has been set to any",
+			serviceMonitor.Name, r.Config.GetNamespace())
+		return nil
+	}
+	sm := &v1.ServiceMonitor{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      serviceMonitor.Namespace + `-` + serviceMonitor.Name,
+			Namespace: r.Config.GetNamespace(),
+		},
+	}
+	opRes, err := controllerutil.CreateOrUpdate(ctx, serverClient, sm, func() error {
+		// Check if the servicemonitor has no  namespace selectors defined,
+		// if not add the namespace
+		sm.Spec = serviceMonitor.Spec
+		if len(sm.Spec.NamespaceSelector.MatchNames) == 0 {
+			sm.Spec.NamespaceSelector.MatchNames = []string{serviceMonitor.Namespace}
+		}
+		//Add all the original labels and append cloned servicemonitor label
+		sm.Labels = serviceMonitor.Labels
+		if len(sm.Labels) == 0 {
+			sm.Labels = make(map[string]string)
+		}
+		sm.Labels[clonedServiceMonitorLabelKey] = clonedServiceMonitorLabelValue
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	r.Logger.Infof("operation result of creating servicemonitor %v was %v", sm.Name, opRes)
+
+	//Get the service monitor - that was created/updated
+	sermon := &v1.ServiceMonitor{}
+	err = serverClient.Get(ctx, k8sclient.ObjectKey{Name: sm.Name, Namespace: r.Config.GetNamespace()}, sermon)
+	if err != nil {
+		return err
+	}
+	//Create role binding for each of the namespace label selectors
+	for _, namespace := range sermon.Spec.NamespaceSelector.MatchNames {
+		err := r.reconcileRoleBindings(ctx, serverClient, namespace)
+		if err != nil {
+			return err
+		}
+	}
+	return err
+}
+
+func (r *Reconciler) reconcileRoleBindings(ctx context.Context,
+	serverClient k8sclient.Client, namespace string) (err error) {
+
+	roleBinding := &rbac.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      roleBindingName,
+			Namespace: namespace,
+		},
+	}
+	opRes, err := controllerutil.CreateOrUpdate(ctx, serverClient, roleBinding, func() error {
+		roleBinding.Subjects = []rbac.Subject{
+			{
+				Kind:      rbac.ServiceAccountKind,
+				Name:      clusterMonitoringPrometheusServiceAccount,
+				Namespace: clusterMonitoringNamespace,
+			},
+		}
+		roleBinding.RoleRef = rbac.RoleRef{
+			APIGroup: roleRefAPIGroup,
+			Kind:     bundle.ClusterRoleKind,
+			Name:     roleRefName,
+		}
+		return nil
+	})
+	r.Logger.Infof("operation result of creating rolebinding: %v was %v", roleBindingName, opRes)
+	return err
+}
+
+func (r *Reconciler) removeServicemonitor(ctx context.Context,
+	serverClient k8sclient.Client, namespace, name string) (err error) {
+	//Get the service monitor
+	sm := &v1.ServiceMonitor{}
+	err = serverClient.Get(ctx, k8sclient.ObjectKey{Name: name, Namespace: namespace}, sm)
+	if err != nil && k8serr.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	//Delete the servicemonitor
+	err = serverClient.Delete(ctx, sm)
```

Comment:
```suggestion
	sm := &v1.ServiceMonitor{ObjectMeta: {Name: name, Namespace: namespace}}
	err = serverClient.Delete(ctx, sm)
```

I don't think it's necessary to do a Get on the ServiceMonitor just so that it can be deleted. You're checking the error for an `IsNotFound` case on the Delete call anyway.

Reply:
Agreed and fixed.